internal/lvmd/command/lvm_lv_attr.go

```diff
@@ -0,0 +1,282 @@
+	VolumeHealthThinFailed               = 'F'
+	VolumeHealthThinPoolOutOfDataSpace   = 'D'
+	VolumeHealthThinPoolMetadataReadOnly = 'M'
+	VolumeHealthWriteCacheError          = 'E'
+)
+
+type SkipActivation rune
+
+const (
+	SkipActivationTrue  SkipActivation = 'k'
+	SkipActivationFalse SkipActivation = '-'
+)
+
+// LvAttr has mapped lv_attr information, see https://linux.die.net/man/8/lvs
+// It is a complete parsing of the entire attribute byte flags that is attached to each LV.
+// This is useful when attaching logic to the state of an LV as the state of an LV can be determined
+// from the Attributes, e.g. for determining whether an LV is considered a Thin-Pool or not.
+type LvAttr struct {
+	VolumeType
+	Permissions
+	AllocationPolicy
+	Minor
+	State
+	Open
+	OpenTarget
+	Zero
+	VolumeHealth
+	SkipActivation
+}
+
+const lvAttrLength = 10
+
+func ParsedLvAttr(raw string) (LvAttr, error) {
+	if len(raw) != lvAttrLength {
+		return LvAttr{}, fmt.Errorf("%s is an invalid length lv_attr, expected %v, but got %v",
+			raw, lvAttrLength, len(raw))
+	}
+	return LvAttr{
+		VolumeType(raw[0]),
+		Permissions(raw[1]),
+		AllocationPolicy(raw[2]),
+		Minor(raw[3]),
+		State(raw[4]),
+		Open(raw[5]),
+		OpenTarget(raw[6]),
+		Zero(raw[7]),
+		VolumeHealth(raw[8]),
+		SkipActivation(raw[9]),
+	}, nil
+}
+
+func (l LvAttr) String() string {
+	return fmt.Sprintf(
+		"%c%c%c%c%c%c%c%c%c",
+		l.VolumeType,
+		l.Permissions,
+		l.AllocationPolicy,
+		l.Minor,
+		l.State,
+		l.Open,
+		l.OpenTarget,
+		l.Zero,
+		l.VolumeHealth,
+	)
+}
+
+// VerifyHealth checks the health of the logical volume based on the attributes, mainly
+// bit 9 (volume health indicator) based on bit 1 (volume type indicator)
+// All failed known states are reported with an error message.
+func (l LvAttr) VerifyHealth() error {
+	if l.VolumeHealth == VolumeHealthPartialActivation {
+		return ErrPartialActivation
+	}
+	if l.VolumeHealth == VolumeHealthUnknown {
+		return ErrUnknownVolumeHealth
+	}
+	if l.VolumeHealth == VolumeHealthWriteCacheError {
+		return ErrWriteCacheError
+	}
+
+	if l.VolumeType == VolumeTypeThinPool {
+		switch l.VolumeHealth {
+		case VolumeHealthThinFailed:
+			return ErrThinPoolFailed
+		case VolumeHealthThinPoolOutOfDataSpace:
+			return ErrThinPoolOutOfDataSpace
+		case VolumeHealthThinPoolMetadataReadOnly:
+			return ErrThinPoolMetadataReadOnly
+		}
+	}
+
+	if l.VolumeType == VolumeTypeThinVolume {
+		switch l.VolumeHealth {
+		case VolumeHealthThinFailed:
+			return ErrThinVolumeFailed
+		}
+	}
+
+	if l.VolumeType == VolumeTypeRAID || l.VolumeType == VolumeTypeRAIDNoInitialSync {
+		switch l.VolumeHealth {
+		case VolumeHealthRAIDRefreshNeeded:
+			return ErrRAIDRefreshNeeded
+		case VolumeHealthRAIDMismatchesExist:
+			return ErrRAIDMismatchesExist
+		case VolumeHealthRAIDReshaping:
+			return ErrRAIDReshaping
+		case VolumeHealthRAIDReshapeRemoved:
+			return ErrRAIDReshapeRemoved
+		case VolumeHealthRAIDWriteMostly:
+			return ErrRAIDWriteMostly
```

Comment:
I'm not sure if the "writemostly" should be reported as an error. It can be changed by `lvchange --writemostly`.

Reply:
I think the explanation of the error should be considered:

```
RAID volume is marked as write-mostly. this signifies the devices in a RAID 1 logical volume have been marked write-mostly. This means that reading from this device will be avoided, and other devices will be preferred for reading (unless no other devices are available). This minimizes the I/O to the specified device
```

I think that just because its easy to change a RAID volume into and away from write-mostly doesn't mean we shouldn't signal I/O degradation. WDYT?

Reply:
Thanks!

I reconsidered the write-mostly, and now I have come to think that it is good to emit an event when the write-mostly flag is found. The current message is simple and easy to understand to me.

I'll close this comment.